### PR TITLE
Prepare mod_disco for dynamic domains

### PR DESCRIPTION
**Pull request by chrzaszcz**
Untangle the service discovery logic according to the following rules:
- No modules should call functions from`mod_disco`. In general, a module shouldn't call other extension modules unless they are connected with `deps`.
- All extra features need to be added using the hook handlers, not with `register_feature`.
- `mod_disco` shouldn't keep its state in ETS tables. There is no need for state when the hooks are used.

Common logic for handling the hooks is extracted to `mongoose_disco`. It has helpers for `disco_local_features` and `disco_local_items` right now.

A lot of new tests for service discovery are added as well.

**To do**:
- Finish the rework, making `disco_local_identity` and `disco_sm_*` consistent with the functions already changed. 
- Support host types in `mod_disco` and `mongoose_disco`.

Both changes can be done in one PR.

### CODE REVIEW

**Review by DenysGonchar — APPROVED** [review]
I like these changes, special thanks for SUITEs extension.

File: big_tests/tests/adhoc_SUITE.erl [hunk 1]
```diff
@@ -50,20 +58,73 @@ init_per_testcase(CaseName, Config) ->
...
-                                      Host)),
-
-                      %% Server replies to Alice with pong
-                      AdHocResp = escalus_client:wait_for_stanza(Alice),
-                      escalus:assert(is_adhoc_response, [<<"ping">>, <<"completed">>],
-                                     AdHocResp)
-                  end).
+    escalus:fresh_story(Config, [{alice, 1}],
+        fun(Alice) ->
+                Host = ct:get_config({hosts, mim, domain}),
```
Comment (arcusfelis) [inline big_tests/tests/adhoc_SUITE.erl:118]: there is domain() defined below

File: big_tests/tests/disco_and_caps_SUITE.erl [hunk 2]
```diff
@@ -54,3 +56,16 @@ user_can_query_server_caps_via_disco(Config) ->
...
     escalus_connection:stop(Alice).
+
+extra_domains_are_advertised(Config) ->
+    escalus:fresh_story(Config, [{alice, 1}], fun(Alice) ->
+        Server = escalus_client:server(Alice),
+        escalus:send(Alice, escalus_stanza:service_discovery(Server)),
+        Stanza = escalus:wait_for_stanza(Alice),
+        escalus:assert(has_service, [extra_domain()], Stanza),
+        escalus:assert(is_stanza_from,
+                            [ct:get_config({hosts, mim, domain})], Stanza)
```
Comment (arcusfelis) [inline big_tests/tests/disco_and_caps_SUITE.erl:67]: same, domain() declared above

File: src/mam/mod_mam.erl [hunk 3]
```diff
@@ -688,7 +695,8 @@ config_metrics(HostType) ->
 
 -spec hooks(jid:lserver()) -> [ejabberd_hooks:hook()].
 hooks(HostType) ->
-    [{user_send_packet, HostType, ?MODULE, user_send_packet, 60},
+    [{disco_local_features, HostType, ?MODULE, add_local_features, 99},
```
Comment (arcusfelis) [inline src/mam/mod_mam.erl:698]: maybe `{disco_local_features, HostType, ?MODULE, disco_local_features, 99}`

Comment (chrzaszcz) [reply src/mam/mod_mam.erl:698]: This name would be consistent with the hook, but on the other hand `add_local_features` is more descriptive as the features are added to the list... Anyway, I can make the handler names more consistent in part 2.

**Comment by DenysGonchar** [comment]
still LGTM

**Review by arcusfelis — APPROVED** [review]

File: src/mod_extdisco.erl [hunk 4]
```diff
@@ -31,14 +31,12 @@
...
     IQDisc = gen_mod:get_opt(iqdisc, Opts, no_queue),
-    mod_disco:register_feature(Host, ?NS_EXTDISCO),
     gen_iq_handler:add_iq_handler(ejabberd_local, Host,
         ?NS_EXTDISCO, ?MODULE, process_iq, IQDisc).
 
 -spec stop(jid:server()) -> ok.
 stop(Host) ->
-    mod_disco:unregister_feature(Host, ?NS_EXTDISCO),
-    gen_iq_handler:remove_iq_handler(ejabberd_sm, Host, ?NS_EXTDISCO).
+    gen_iq_handler:remove_iq_handler(ejabberd_local, Host, ?NS_EXTDISCO).
```
Comment (NelsonVides) [inline src/mod_extdisco.erl:39]: Good one!

**Review by NelsonVides — APPROVED** [review]
Love these changes! 😄

**Pull request merged**